java/src/main/java/com/adhoc/flight/client/AdhocFlightClient.java

```diff
@@ -336,13 +363,27 @@ public void runQuery(final String query,
                        final @Nullable File fileToSaveTo,
                        final boolean printToConsole) throws Exception {
 
+    if (projectId != null) {
+      SetSessionOptionsResult res1 = client.setSessionOptions(
+          createSetSessionOption(PROJECT_ID_KEY, projectId), bearerToken, headerCallOption);
+      if (res1.hasErrors()) {
+        Map<String, SetSessionOptionsResult.Error> errors = res1.getErrors();
+        for (Entry<String, Error> error : errors.entrySet()) {
+          System.out.println(error.toString());
+        }
+      }
+    }
     final FlightInfo flightInfo = getInfo(query, bearerToken, headerCallOption);
     try (final FlightStream flightStream = getStream(flightInfo, bearerToken, headerCallOption);
          final OutputStream outputStream =
-             fileToSaveTo == null ? null : new BufferedOutputStream(new FileOutputStream(fileToSaveTo))) {
+             fileToSaveTo == null ? null : new BufferedOutputStream(
+                 Files.newOutputStream(fileToSaveTo.toPath()))) {
       writeToOutputStream(
           flightStream, allocator, outputStream, printToConsole ? QueryUtils::printResults : null);
     }
+    if (projectId != null) {
+      client.closeSession(new CloseSessionRequest(), bearerToken, headerCallOption);
```

Comment:
Is it better to use try with finally? Should we close the session with an empty projectId as well?

Reply:
No, if setSessionOptions has errors a session will not be created.
We won't create a session unless there is an option to add. A session doesn't have meaning without options.